[Reverted] Module-builtin assertions, disabling assertions and submodule assertions

infinisil opened the pull request:
### Motivation for this change
This implements checks (assertions and warnings) supported by the module system directly, instead of just being a NixOS option (see [nixos/modules/misc/assertions.nix](https://github.com/NixOS/nixpkgs/blob/53bce4da36f7a789929c983d29aebbaf873c7bef/nixos/modules/misc/assertions.nix)).

This has the following benefits:
- It allows cleanly redoing the user interface. The new implementation specifically allows:
  - disabling checks
  - converting errors to warnings
- Checks can now be specified easily from within submodules, which previously wasn't possible and needed workarounds
  - In particular this also allows using `mkRemovedOptionModule` and co. from within submodules

#### Usage
Checks can now be defined in any module (even submodules) like this:
```nix
{ config, options, ... }: {
  _module.checks.gpgSshAgent = {
    check = config.programs.gnupg.agent.enableSSHSupport -> !config.programs.ssh.startAgent;
    message = "If you have ${options.programs.gnupg.agent.enableSSHSupport} enabled,"
      + " you can't enable ${options.programs.ssh.startAgent} as well!";
  };

  _module.checks.grafanaPassword = {
    check = config.services.grafana.database.password == "";
    message = "The grafana password defined with ${options.services.grafana.database.password}"
      + " will be stored as plaintext in the Nix store!";
    # This is a non-fatal warning
    type = "warning";
  };
}
```

Triggered assertion will display the attribute name they were defined as in `[..]`:
```
trace: warning: [grafanaPassword] The grafana password defined with
  services.grafana.database.password will be stored as plaintext in the Nix store!
error: Failed checks:
- [gpgSshAgent] If you have programs.gnupg.agent.enableSSHSupport
  enabled, you can't enable programs.ssh.startAgent as well!
```

This allows you to easily disable them or convert an assertion to a warning by using the same attribute that was printed:
```nix
{
  # Change the assertion into a non-fatal warning
  _module.checks.gpgSshAgent.type = "warning";

  # We don't care about this warning, disable it
  _module.checks.grafanaPassword.enable = false;
}
```

See the manual changes for more details (such as how submodule assertions work). Here is a screenshot of it:
<details>

![Screen Shot 2020-11-30 at 23 59 31](https://user-images.githubusercontent.com/20525370/100675917-25e03c80-3368-11eb-9306-e9a92fc3f47c.png)
</details>

Fixes https://github.com/NixOS/nixpkgs/issues/96006

Ping @roberth @nbp @shlevy @edolstra @jD91mZM2 @aanderse 

Best reviewed commit-by-commit

Note that in order to be able to change checks (disable or convert to warning), they _need_ to be defined using `_module.checks`, such that they have a name you can reference. Checks defined using the NixOS options `assertions` or `warnings` can't be changed (even though they're implemented using `_module.checks` now). While the intention is that all of the assertions/warnings in NixOS be given a name and be converted to `_module.checks` eventually, this PR doesn't do that so it stays manageable in size.

### Things done

- [x] Wrote tests
- [x] Wrote docs
- [x] Tested performance (no measurable difference for my machines configs)

infinisil committed:
lib/modules: Implement module-builtin assertions

This implements assertions/warnings supported by the module system directly,
instead of just being a NixOS option (see
nixos/modules/misc/assertions.nix).

This has the following benefits:
- It allows cleanly redoing the user interface. The new
  implementation specifically allows disabling assertions or
  converting them to warnings instead.
- Assertions/warnings can now be thrown easily from within
  submodules, which previously wasn't possible and needed workarounds.

infinisil committed:
nixos/assertions: Use module-builtin assertion implementation

infinisil committed:
lib/options: Don't show internal suboption in the manual

Initially https://github.com/NixOS/nixpkgs/pull/82897 prevented
non-visible options from being rendered in the manual, but
visible-but-internal options were still being recursed into. This fixes
this, aligning the recurse condition here with the one in
make-options-doc/default.nix

infinisil committed:
lib/modules: Use module-builtin assertions for mkRemovedOptionModule and co.

infinisil committed:
nixos/modules: Allow options to be coerced to a string for convenience

infinisil committed:
nixos/modules: Expose the internal module in the top-level documentation

infinisil committed:
nixos/docs: Update assertion docs for new module-builtin ones

infinisil committed:
lib/tests: Implement generalized checkConfigCodeOutErr for module tests

infinisil committed:
lib/tests: Add tests for module-builtin assertions

bqv commented:
Awesome stuff!

roberth commented:
Perhaps we could add syntax sugar to make it top-level like `config`, `freeformType` etc? I don't think we should recommend defining stuff in `_module`. This will be everywhere.

It'd be great to have a way to turn all warnings into errors too.

I'll have to take a closer look at `triggerPath`, but do you think it's still going to be possible to collect all assertions and warnings without relying on `trace` and `throw`? I suppose even if we can, some configurations may turn out to accidentally(?) rely on laziness to avoid some checks.

Which reminds me - `checks`, perhaps that's usable as a generalization of assertions and warnings? For that matter, even derivations can serve as checks. So you'd have `{config, ...}: { checks.portConflict.type = "assertion"; checks.checkConfigSyntax.check = pkgs.runCommandNoCC .....; }`. It does collide somewhat with flake `checks` though, terminology-wise. Not sure if too out-of-the-box, but it does affect the naming scheme, so it's worth considering.

Infinisil commented:
> Perhaps we could add syntax sugar to make it top-level like `config`, `freeformType` etc? I don't think we should recommend defining stuff in `_module`. This will be everywhere.

This is a bit tricky because there's already the `assertions` NixOS option. Maybe the sugar could be `moduleAssertions` instead though.

> It'd be great to have a way to turn all warnings into errors too.

While I don't think this is a good idea, this can already be done with
```nix
{
  options._module.assertions = mkOption {
    type = attrsOf (submodule {
      type = mkDefault "warning";
    });
  };
}
```

> I'll have to take a closer look at `triggerPath`, but do you think it's still going to be possible to collect all assertions and warnings without relying on `trace` and `throw`? I suppose even if we can, some configurations may turn out to accidentally(?) rely on laziness to avoid some checks.

I played around with them being collected on a per-module eval basis, such that they could be forwarded from submodules to the main module eval. I might give this another go.

> Which reminds me - `checks`, perhaps that's usable as a generalization of assertions and warnings? For that matter, even derivations can serve as checks. So you'd have `{config, ...}: { checks.portConflict.type = "assertion"; checks.checkConfigSyntax.check = pkgs.runCommandNoCC .....; }`. It does collide somewhat with flake `checks` though, terminology-wise. Not sure if too out-of-the-box, but it does affect the naming scheme, so it's worth considering.

I like calling this `checks` instead. Interesting idea, I'll think about it.

Profpatsch commented:
> This is a bit tricky because there's already the `assertions` NixOS option. Maybe the sugar could be `moduleAssertions` instead though.

I think he meant on the same level as `config` `options` and `imports`.

Infinisil commented:
@Profpatsch The problem is that if a module is just `{ assertions = ...; }` this becomes ambiguous: Is it the NixOS option `assertions` or is it the module-builtin `assertions`? I'd rather avoid the confusion and explicitly call it something different

bqv commented:
+1 for checks

roberth commented:
> The problem is that if a module is just `{ assertions = ...; }` this becomes ambiguous: Is it the NixOS option `assertions` or is it the module-builtin `assertions`? I'd rather avoid the confusion and explicitly call it something different

The shorthand syntax is fundamentally ambiguous, so the same can be argued about `imports`.

We don't have any `checks` options at the root of any submodules so I don't expect this to be a problem in practice.

Profpatsch commented:
> The problem is that if a module is just `{ assertions = ...; }` this becomes ambiguous:

Do we have any modules that are only assertions? In that case we can migrate it manually; it seems low-chance enough that we can do a breaking change here (provided we have a good error message and migration documentation in the NixOS release notes).

Is it currently possible to define a module that has only e.g. `options` defined? I was under the impression that it will check for `config`s existence and based on that use the shortcut or not.

infinisil committed:
lib/modules: Rename _module.assertions to _module.checks

roberth commented:
Part of this feature could be a generalized facility for values that propagate to the root of the configuration monoidally, not unlike a writer monad when used in a tree traversal. That's also how assertions and warnings should propagate, so the only extra bits of code you need are the `assert` and `trace` calls near `evalModules` returned `config` attribute.
Such a feature may be one of the ingredients for a radically different style of module writing, where modules stand alone like functions and somewhat more naturally solve the infamous "multiple database services" problem. It could be similar to an effect + handler system.

If this resonates with you, great! Otherwise feel free to ignore.

Infinisil commented:
@roberth I kinda get it, but I think that's a good discussion for another time :smiley: 

Regarding introducing syntactic sugar, I thought long and hard about this now, and am coming to the conclusion that we should not do it.

The main insight I had was that `_module` doesn't have a `_` at the start because it's for module-internal things, but rather to avoid clashing with option names. While @roberth pointed out that we don't have a `checks` option at the root, these options are added to submodules as well. So introducing sugar for this could cause problems for any module or submodule which has a `checks` option. While there is no such option in nixpkgs, we can't be sure that users don't have one on their own.

In addition, syntactic sugar would interact in non-optimal ways with submodules: You can't do
```nix
{
  myServices.foo.checks.portConflict.enable = lib.mkForce false;
}
```
because due to how submodules are implemented (for the default of `shorthandOnlyDefinesConfig = true`), this turns into `config.checks = ...`, which is not right, and as far as I can see, there's no reasonable way to make this work nicely. So people would still have to use `_module.checks` in such a case anyways.

---

In other news, it turns out that `triggerPath` isn't needed at all, so I'll remove that.

I'll also do the renaming from `_module.assertions` to `_module.checks`

And I think after that this PR should be about ready.

Infinisil commented:
Also, I'll introduce a `_module.<name>.check` option, which is the condition that needs to hold true for the check to succeed. While this is very similar to `_module.<name>.enable`, having this be separate is easier to reason about, and it allows reenabling the check after it's been turned off with `enable = false`.

infinisil committed:
lib/modules: Remove _module.checks.*.triggerPath as it's not necessary

Previously this option was thought to be necessary to avoid infinite
recursion, but it actually isn't, since the check evaluation isn't fed
back into the module fixed-point.

infinisil committed:
lib/modules: _module.check should always be internal

Honestly this option should probably just be removed

infinisil committed:
lib/modules: Introduce _module.checks.*.check

Previously the .enable option was used to encode the condition as well,
which lead to some oddness:
- In order to encode an assertion, one had to invert it
- To disable a check, one had to mkForce it

By introducing a separate .check option this is solved because:
- It can be used to encode assertions
- Disabling is done separately with .enable option, whose default can be
  overridden without a mkForce

Infinisil commented:
Alright I think this PR is ready. I opted for `_module.checks.<name>.assertion` instead of `.check` or `.condition`, because `.assertion` is clearer that it needs to succeed.

I rebased the commit on top of master, fixing any intermediate issues introduced (due to https://github.com/NixOS/nixpkgs/pull/101139), so the initial 9 commits are pretty much the same as before.

The last 4 commits are new. I also updated the PR description at the top to match the latest state.

@roberth @Profpatsch Would love a (re-)review

Infinisil commented:
Actually, one more thing that should be done here is the addition of tests for `mkRemovedOptionModule` and co. These previously didn't have any tests partially because `assertions`/`warnings` was a NixOS option and not a module-internal one. We can now easily add tests for these. I'll probably do so soon.

roberth commented on lib/modules.nix:
```diff
@@ -116,6 +129,75 @@ rec {
+            type = types.attrsOf (types.submodule {
+              options.enable = mkOption {
+                description = ''
+                  Whether to enable this check.
```
```suggestion
                  A last resort option to disable a check by setting it to <code>false</code>.
```

Infinisil replied:
I don't think "last resort" is fitting here, as you can still enable it again after that (e.g. with `mkForce`). I'll just add an extra sentence along the lines of

> `Set this to <literal>false</literal> to not trigger any errors or warning messages for this check, e.g. in case it doesn't apply for your case.`

roberth commented on lib/modules.nix:
```diff
@@ -116,6 +129,75 @@ rec {
+                type = types.bool;
+              };
+
+              options.assertion = mkOption {
```
```suggestion
              options.check = mkOption {
```

The term `assertion` is incompatible with `type = "warning"`.
Normally an assertion means that a condition _must_ hold, with the effect of program cancellation when it turns out not to.

Infinisil replied:
Yeah I guess that's a good point, I'll change it to `check`

roberth commented on lib/modules.nix:
```diff
@@ -116,6 +129,75 @@ rec {
+                '';
+                type = types.str;
+                example = literalExample ''
+                  Enabling both ''${options.services.foo.enable} and ''${options.services.bar.enable} is not possible.
```
This will render as function application.
```nix
Enabling both ${options.services.foo.enable} and ${options.services.bar.enable} is not possible.
```
instead of a string literal
```nix
"Enabling both ${options.services.foo.enable} and ${options.services.bar.enable} is not possible."
```

Perhaps `literalExample` should have been called `verbatimExample`.

Infinisil commented:
I tested this change myself a bit more, and I couldn't find anything that wasn't working.

Unless there are any complaints, or unless somebody is quicker than me, I'll merge this soon :)

Infinisil commented:
@roberth @Profpatsch (Gentle ping)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-5/10560/1

roberth commented on lib/modules.nix:
```diff
@@ -773,14 +886,15 @@ rec {
-          assertion = !opt.isDefined;
+      config._module.checks =
+        let opt = getAttrFromPath optionName options; in {
+        ${showOption optionName} = lib.mkIf opt.isDefined {
```
```suggestion
        ${"removed-" + showOption optionName} = lib.mkIf opt.isDefined {
```

Avoid a bad, unexpected error message in case someone named a domain specific check the same as an option they're removing.

roberth commented on lib/modules.nix:
```diff
@@ -841,14 +955,18 @@ rec {
+            let val = getAttrFromPath f config;
+                opt = getAttrFromPath f options;
+            in {
+              ${showOption f} = lib.mkIf (val != "_mkMergedOptionModule") {
```
```suggestion
              ${"merged-" + showOption f} = lib.mkIf (val != "_mkMergedOptionModule") {
```

roberth commented on lib/modules.nix:
```diff
@@ -907,8 +1025,10 @@ rec {
         {
-          warnings = optional (warn && fromOpt.isDefined)
-            "The option `${showOption from}' defined in ${showFiles fromOpt.files} has been renamed to `${showOption to}'.";
+          _module.checks.${showOption from} = mkIf (warn && fromOpt.isDefined) {
```
```suggestion
          _module.checks.${"renamed-" + showOption from} = mkIf (warn && fromOpt.isDefined) {
```
same

roberth reviewed (commented):
One last thing, which is the potential for name clashes.
Should be good for merge after this.
Do you work for tweag now? I should start charging ;)

Infinisil commented on lib/modules.nix:
```diff
@@ -116,6 +129,75 @@ rec {
+                  The condition that must succeed in order for this check to be
+                  successful and not trigger a warning or error.
+                '';
+                default = false;
```
`readOnly` sounds good. If people wanted to change a `check` condition, they should rather disable the one they don't want and define a new one instead (with a potentially better message). If the need for read-write later arises we could still remove it though.

infinisil committed:
lib/modules: Prefix mkRemovedOptionModule & co. check names

To avoid name clashes

Co-authored-by: Robert Hensing <robert@roberthensing.nl>

Infinisil commented:
@roberth Addressed the comments. And indeed, I'm working for Tweag now :wink:

roberth commented:
Cool. lgtm.

roberth commented:
idk why ofborg hasn't run yet

Infinisil commented:
@GrahamcOfBorg eval

Infinisil commented:
Seems borged up (hah). I'll go for the merge, thanks for the reviews!

NeQuissimus commented:
@Infinisil Somehow, this introduces infinite recursions:

```
λ nix-build -A minecraft.tests --show-trace
error: while evaluating the attribute 'buildCommand' of the derivation 'vm-test-run-minecraft' at /opt/dev/upstream_nix/nixos/lib/testing-python.nix:60:7:
while evaluating the attribute 'buildCommand' of the derivation 'nixos-test-driver-minecraft' at /opt/dev/upstream_nix/pkgs/build-support/trivial-builders.nix:7:7:
while evaluating anonymous function at /opt/dev/upstream_nix/nixos/lib/testing-python.nix:119:22, called from undefined position:
while evaluating the attribute 'config.system.build.vm' at /opt/dev/upstream_nix/nixos/lib/eval-config.nix:64:5:
while evaluating the attribute 'config' at /opt/dev/upstream_nix/lib/modules.nix:287:9:
while evaluating 'handleCheck' at /opt/dev/upstream_nix/lib/modules.nix:242:31, called from /opt/dev/upstream_nix/lib/modules.nix:255:18:
while evaluating the attribute 'check' at undefined position:
while evaluating 'g' at /opt/dev/upstream_nix/lib/attrsets.nix:276:19, called from undefined position:
while evaluating anonymous function at /opt/dev/upstream_nix/lib/modules.nix:221:72, called from /opt/dev/upstream_nix/lib/attrsets.nix:279:20:
while evaluating the attribute 'value' at /opt/dev/upstream_nix/lib/modules.nix:623:9:
while evaluating the option `_module.checks._28.check':
while evaluating the attribute 'mergedValue' at /opt/dev/upstream_nix/lib/modules.nix:657:5:
while evaluating the attribute 'values' at /opt/dev/upstream_nix/lib/modules.nix:651:9:
while evaluating the attribute 'values' at /opt/dev/upstream_nix/lib/modules.nix:750:7:
while evaluating anonymous function at /opt/dev/upstream_nix/lib/modules.nix:637:28, called from /opt/dev/upstream_nix/lib/modules.nix:637:17:
while evaluating definitions from `/opt/dev/upstream_nix/nixos/modules/misc/assertions.nix':
while evaluating 'dischargeProperties' at /opt/dev/upstream_nix/lib/modules.nix:709:25, called from /opt/dev/upstream_nix/lib/modules.nix:638:137:
while evaluating the attribute 'value' at /opt/dev/upstream_nix/lib/modules.nix:490:44:
while evaluating the attribute 'assertion' at /opt/dev/upstream_nix/nixos/modules/services/x11/display-managers/default.nix:369:9:
while evaluating the attribute 'displayManager.autoLogin.user' at undefined position:
while evaluating 'g' at /opt/dev/upstream_nix/lib/attrsets.nix:276:19, called from undefined position:
while evaluating anonymous function at /opt/dev/upstream_nix/lib/modules.nix:221:72, called from /opt/dev/upstream_nix/lib/attrsets.nix:279:20:
while evaluating the attribute 'value' at /opt/dev/upstream_nix/lib/modules.nix:623:9:
while evaluating the option `services.xserver.displayManager.autoLogin.user':
while evaluating the attribute 'mergedValue' at /opt/dev/upstream_nix/lib/modules.nix:657:5:
while evaluating the attribute 'values' at /opt/dev/upstream_nix/lib/modules.nix:651:9:
while evaluating the attribute 'values' at /opt/dev/upstream_nix/lib/modules.nix:750:7:
while evaluating anonymous function at /opt/dev/upstream_nix/lib/modules.nix:637:28, called from /opt/dev/upstream_nix/lib/modules.nix:637:17:
while evaluating definitions from `/opt/dev/upstream_nix/nixos/tests/common/auto.nix':
while evaluating 'dischargeProperties' at /opt/dev/upstream_nix/lib/modules.nix:709:25, called from /opt/dev/upstream_nix/lib/modules.nix:638:137:
while evaluating the attribute 'value' at /opt/dev/upstream_nix/lib/modules.nix:490:44:
while evaluating the attribute 'user' at undefined position:
while evaluating 'g' at /opt/dev/upstream_nix/lib/attrsets.nix:276:19, called from undefined position:
while evaluating anonymous function at /opt/dev/upstream_nix/lib/modules.nix:221:72, called from /opt/dev/upstream_nix/lib/attrsets.nix:279:20:
while evaluating the attribute 'value' at /opt/dev/upstream_nix/lib/modules.nix:623:9:
while evaluating the option `test-support.displayManager.auto.user':
while evaluating the attribute 'mergedValue' at /opt/dev/upstream_nix/lib/modules.nix:657:5:
while evaluating the attribute 'values' at /opt/dev/upstream_nix/lib/modules.nix:651:9:
while evaluating the attribute 'values' at /opt/dev/upstream_nix/lib/modules.nix:750:7:
while evaluating anonymous function at /opt/dev/upstream_nix/lib/modules.nix:637:28, called from /opt/dev/upstream_nix/lib/modules.nix:637:17:
while evaluating definitions from `<unknown-file>':
while evaluating 'dischargeProperties' at /opt/dev/upstream_nix/lib/modules.nix:709:25, called from /opt/dev/upstream_nix/lib/modules.nix:638:137:
while evaluating the attribute 'value' at /opt/dev/upstream_nix/lib/modules.nix:490:44:
while evaluating the attribute 'client.config.users.users.alice' at /opt/dev/upstream_nix/nixos/lib/eval-config.nix:64:5:
infinite recursion encountered, at undefined position
```

Infinisil commented:
Thanks for the report! I'll try to see if it's a quick fix, otherwise I'll revert this change

NeQuissimus commented:
Sorry, I just noticed this and haven't had much luck figuring this out... Hence the plain log :D

roberth commented:
If you can move the `handleCheck` call outside of the `config` in `{config,...}:` and only add it as the very very last thing in `evalModules`, that might work. Otherwise I think we'll have to revert.

roberth commented:
It already looks like what I was thinking of. At this time I don't have any other suggestions than reverting.

Infinisil commented:
Well, turns out that `triggerPath` might be needed after all, as it doesn't give infinite recursion before https://github.com/NixOS/nixpkgs/pull/97023/commits/8dea4df90323c43f9cc86a629f1581b91866e11d!

Infinisil commented:
I'll revert for now

Infinisil commented:
Will soon open another PR for this

Infinisil commented:
Looking at how the [minecraft test is defined](https://github.com/NixOS/nixpkgs/blob/9e6737710c4fb2613850e699178b23d54f1a3261/nixos/tests/minecraft.nix#L5-L15), it makes sense:
```nix
{
  nodes.client = { nodes, ... }:
      let user = nodes.client.config.users.users.alice;
      in {
        imports = [ ./common/user-account.nix ./common/x11.nix ];

        environment.systemPackages = [ pkgs.minecraft ];

        nixpkgs.config.allowUnfree = true;

        test-support.displayManager.auto.user = user.name;
      };
}
```

`nodes` are defined to be the final system config of all defined nodes, [here](https://github.com/NixOS/nixpkgs/blob/9e6737710c4fb2613850e699178b23d54f1a3261/nixos/lib/build-vms.nix#L25-L41). The problem is that the node references its own config through the `nodes` value, which with the latest state of this PR triggers assertions, which of course depend on the nodes config itself, therefore the infinite recursion.

A `triggerPath` will alleviate the problem since it would only trigger assertions at accessing of e.g. `system.build.toplevel`

roberth commented:
The effect of this PR has made clear that we can't perform checks locally (entirely within small parts of the configuration), because it introduces strictness.

Instead, what we need is to have two global "phases". In the first phase, module evaluation happens exactly as it does now. It must not and can not depend on the checks. In the second phase, the checks are performed. This may still evaluate some parts that weren't evaluated during the first, but as said, won't refer back to the second phase to cause trouble.

While we can't enforce this in arbitrary Nix expressions, we can make this easy with the module system.
The first phase can be represented by a new `(evalModules x).lazyConfig`. The second will be represented by `(evalModules x).config`. The latter is defined as something like `builtins.seq (runChecks lazyConfig._module.checks) lazyConfig`, so nothing is duplicated.

Now it's impossible for modules/submodules to cause the problem, but we haven't actually wired up the checks from the submodules yet. Even without this, the change is already a positive because it standardizes the interface for writing checks and running checks without causing strictness issues.

Module authors could wire it up manually with something like

```
{
  config = mkIf cfg.enable {
    _module.checks = f options.services.foo.virtualHosts._module.checks;
  };
}
```

if we expose `_module` in `options`; similar to `value`, selecting `_module` instead of removing it.

Alternatively or additionally, we could automate this by scanning all options for submodules.


> `nodes` are defined to be the final system config of all defined nodes, [here](https://github.com/NixOS/nixpkgs/blob/9e6737710c4fb2613850e699178b23d54f1a3261/nixos/lib/build-vms.nix#L25-L41). The problem is that the node references its own config through the `nodes` value, which with the latest state of this PR triggers assertions, which of course depend on the nodes config itself, therefore the infinite recursion.

This won't be a problem with the above proposal, if the nodes are defined through `attrsOf submodule`.

If that's not feasible (I'd be surprised) the code could just use `lazyConfig` and do the `seq foo.config` on its own.

> A `triggerPath` will alleviate the problem since it would only trigger assertions at accessing of e.g. `system.build.toplevel`

This can still cause strictness issues that then can't be avoided. Assertions won't be able to use `toplevel`, which is not top level anyway. Image modules use it and you could reverse the toplevel and activation dependency for immutable systems. Another potential problem is that you could evaluate a configuration for only some options that do not include `toplevel` but should be checked nonetheless.

roberth commented:
This seems to be one of the few 'architectural' weaknesses that can't be addressed by good module design.

roberth commented:
Supermodules (#152785) are another use case for an out-of-band (non-`config`) module return value.

Infinisil commented:
I want to look into this again. For now I just rebased this PR on top of latest master to evaluate it again, here's that for reference: https://github.com/Infinisil/nixpkgs/tree/module-assertions-v2

roberth commented:
I think what I'd try is

1. Extend types and `options` so we can retrieve all submodules out-of-band,
2. Write a function for inheriting checks from an option,
3. Rename `evalModules` to `evalModulesLazy` in all of `lib/`,
4. Write `evalModules` using `evalModulesLazy`, but performing the checks before returning `config`. This solves the strictness problem by removing all intermediate "seqs", while module system users still run checks by default.

Making (2) automatic would be nice but also seems like a performance hazard, especially without a minimal module list. It'd be a one-liner inside the `mkIf cfg.enable`. Something like

    config = mkIf cfg.enable {
      checks = inheritChecks options.services.foo;
    };

or maybe it should be like this

    config = mkIf cfg.enable { ..... };
    checks = inheritChecks cfg.enable options.services.foo;

roberth commented:
Not having this is a layering violation, as `lib.mkRenamedOptionModule` and friends reference these options, that are implemented by NixOS and not exposed separately.

Encountered this again while working on the NixOS test framework https://github.com/NixOS/nixpkgs/pull/176557.

infinisil commented:
A discussion relating to this PR is happening here: https://github.com/NixOS/nixpkgs/pull/207187?notification_referrer_id=NT_kwDOATkxOrM1MTUyMTE2MDUyOjIwNTI1Mzcw#discussion_r1055695038
